core/trino-main/src/main/java/io/trino/operator/aggregation/OrderedWindowAccumulator.java

```diff
@@ -0,0 +1,145 @@
+            PagesIndex pagesIndex,
+            WindowAccumulator delegate,
+            List<Type> argumentTypes,
+            List<Integer> argumentChannels,
+            List<SortOrder> sortOrders)
+    {
+        this.pagesIndexFactory = requireNonNull(pagesIndexFactory, "pagesIndexFactory is null");
+        this.pagesIndex = requireNonNull(pagesIndex, "pagesIndex is null");
+        this.argumentTypes = requireNonNull(argumentTypes, "argumentTypes is null");
+        this.argumentChannels = ImmutableList.copyOf(argumentChannels);
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.sortChannels = IntStream.range(argumentTypes.size() - sortOrders.size(), argumentChannels.size())
+                .boxed()
+                .toList();
+
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.initialDelegate = delegate.copy();
+
+        resetPageBuilder();
+    }
+
+    private void resetPageBuilder()
+    {
+        pageBuilder = new PageBuilder(argumentTypes);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return delegate.getEstimatedSize() + initialDelegate.getEstimatedSize() + pagesIndex.getEstimatedSize().toBytes() + pageBuilder.getRetainedSizeInBytes();
+    }
+
+    @Override
+    public WindowAccumulator copy()
+    {
+        PagesIndex pagesIndexCopy = pagesIndexFactory.newPagesIndex(argumentTypes, pagesIndex.getPositionCount());
+        pagesIndex.getPages().forEachRemaining(pagesIndexCopy::addPage);
+        return new OrderedWindowAccumulator(pagesIndexFactory, pagesIndexCopy, delegate.copy(), argumentTypes, argumentChannels, sortOrders);
+    }
+
+    @Override
+    public void addInput(WindowIndex index, int startPosition, int endPosition)
+    {
+        if (pagesIndexSorted) {
+            pagesIndexSorted = false;
+            // operate on delegate as of start
+            // nicer would be to add reset() method to WindowAccumulator but it requires reset method in each AccumulatorState class
+            delegate = initialDelegate.copy();
+        }
+        // index is remapped so just go from 0 to argumentChannels.size()
+        for (int position = startPosition; position <= endPosition; position++) {
+            if (pageBuilder.isFull()) {
+                pagesIndex.addPage(pageBuilder.build());
+                resetPageBuilder();
+            }
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                ValueBlock value = index.getSingleValueBlock(channel, position).getSingleValueBlock(0);
+                pageBuilder.getBlockBuilder(channel).append(value, 0);
+            }
+            pageBuilder.declarePosition();
+        }
+    }
+
+    @Override
+    public void output(BlockBuilder blockBuilder)
+    {
+        if (!pagesIndexSorted) {
+            if (!pageBuilder.isEmpty()) {
+                pagesIndex.addPage(pageBuilder.build());
+                resetPageBuilder();
```

Comment:
You could extract this to a method that returns a Page while resetting page builder

Reply:
Extracted
```java
    private void indexCurrentPage()
    {
        pagesIndex.addPage(pageBuilder.build());
        pageBuilder.reset();
    }
```